src/main/java/org/janelia/saalfeldlab/n5/KeyValueAccessLazyReadData.java

```diff
@@ -0,0 +1,112 @@
+package org.janelia.saalfeldlab.n5;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+import org.janelia.saalfeldlab.n5.N5Exception.N5IOException;
+import org.janelia.saalfeldlab.n5.readdata.ReadData;
+
+/**
+ * This abstract class represents a lazy read operation, and implements the
+ * shared logic for lazily reading from a {@link KeyValueAccess}.
+ * 
+ * @param <K>
+ *            the type of {@link KeyValueAccess}.
+ */
+abstract class KeyValueAccessLazyReadData<K extends KeyValueAccess> implements ReadData {
+
+	protected ReadData materialized;
+
+	protected final K kva;
+	protected final String normalKey;
+	protected final long offset;
+	protected long length;
+
+	KeyValueAccessLazyReadData(K kva, String normalKey, long offset, long length) {
+
+		if (offset < 0)
+			throw new IndexOutOfBoundsException("Can not create KeyValueAccesReadData with negative offset: " + offset);
+
+		this.kva = kva;
+		this.normalKey = normalKey;
+		this.offset = offset;
+		this.length = length;
+	}
+
+	KeyValueAccessLazyReadData(K kva, String normalKey, long offset) {
+
+		this(kva, normalKey, offset, -1);
+	}
+
+	@Override
+	public long length() throws N5IOException {
+
+		if (materialized != null)
+			return materialized.length();
+
+		if (length < 0) {
+			length = kva.size(normalKey);
+		}
```

Comment:
probably should be `length = kva.size(normalKey) - offset;` as in the proposal made above